### PR TITLE
[operator] Introduce `Extension` API

### PR DESCRIPTION
**Pull request by kon-angelo**
**How to categorize this PR?**

/area control-plane
/kind enhancement

**What this PR does / why we need it**:

Introduce the extension types for https://github.com/gardener/gardener/issues/9635. Outcome from the recent community hackathon (whose end-result type you can find [here](https://github.com/metal-stack/gardener/blob/hackathon-gardener-operator-extensions-including-complete-local-setup/pkg/apis/operator/v1alpha1/types_extension.go)). Some explanation for the change in fields and a start for discussion:

The type is a combination of the controller-{reg,deploy} fields used to normally install the extensions. 
- The `Resources` field indicates the supported types by the extension.
- The `AdmissionDeployment` contains the charts for the deployment of an admission controller. It is divided into 2 charts the runtime and the garden cluster. This follows the pattern of the extensions e.g. [prov-aws](https://github.com/gardener/gardener-extension-provider-aws/tree/master/charts/gardener-extension-admission-aws/charts)
   - The runtime chart is the actual deployment of the admission controller.
   - The garden cluster handles mainly the installation of RBAC or other resources in the virtual cluster.
   - One alternative would be to support subChart installation and hence rely only on a single chart with different values.
- The `ExtensionDeployment` contains the charts for the deployment of the extension and is divided into 2 parts too similar to the admission. The reason is that a copy of the extension managing infra resources in the runtime cluster may need to differentiate from a typical seed installation. 
   
    The 2 chart approach has the benefit of allowing 2 different versions running between the operator and the seeds. This allows more flexibility with regards to the rollout (separate rollout versions for seeds/operator) or handling the rare case that one issue affects only one environment. **The bare minimum should be 2 different set of values** and a common chart.

There is one field missing when compared to the hackathon branch, which is the `extensionDeployment.annotations`. For the extension deployment in the seeds we have some well-known annotations e.g. for the `podSecurityPolicy`. Instead of introducing new fields, we can use the same concept as the controllerRegistration and read the extension annotations for that.

Also we forgo support for rawCharts and only support OCIRepository types similar to https://github.com/gardener/gardener/pull/9874 

**Which issue(s) this PR fixes**:
Part of #9635

**Special notes for your reviewer**:

**Release note**:

```other operator
Introduce the operator `Extension` type. An `Extension` is responsible for installing the `ControllerRegistration` and `ControllerRuntime` in the garden cluster and also install any extensions necessary to manage infrastructure resources in the runtime cluster. This release contains the API only, further functionality will be released in future Gardener versions.
```

### CODE REVIEW

File: example/operator/15-extension.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+#  admissionDeployment:
+#  extensionDeployment:
+#    runtimeDeployment:
+#      helm:
+#        ociRepository: {}
+#        values: {}
+#    gardenDeployment:
+#      helm:
+#        ociRepository: {}
+#        values: {}
```
Comment (rfranzke) [inline example/operator/15-extension.yaml]: In https://github.com/gardener/gardener/issues/9635, the proposal was to have

```yaml
deployment:
  admission: {}
  extension: {}
```

Why did you deviate from this?

> The reason is that a copy of the extension managing infra resources in the runtime cluster may need to differentiate from a typical seed installation.

Can you explain? Why would an extension (like `provider-aws` or whatever else) need to deploy resources (RBAC, ...) into the virtual cluster? Can you share a scenario/use-case?

Comment (kon-angelo) [reply example/operator/15-extension.yaml]: > Can you explain? Why would an extension (like provider-aws or whatever else) need to deploy resources (RBAC, ...) into the virtual cluster? Can you share a scenario/use-case?

The admission part of each extension which is already neatly separated into 2 charts that target the runtime and garden cluster. The admission needs to deploy it's RBAC to validate/mutate shoot resources, but it is running in the runtime cluster, hence the need for a chart for each of the 2 clusters.

The gardenDeployment section of the `extensionDeployment` refers rather to the `controller{deployment,registration}` part - which is what we actually deploy to the garden cluster. The idea that I tried to convey in the description is that there can be scenarios where the provider-ext running in the operator and the seeds may need to deviate. 
The reasons may be e.g. resource requests needed for the extensions, priority classes and/or other future options, or upgrade/migration scenarios where you want to do a 2 stage upgrade (a stage for the operator version and a stage for the seeds). 

Following that train of though, we need at the very least 2 different values (one for the deployment in operator and one for the `ctrldeploy`. If you think that there will be a need for running the 2 separate versions, then you need the secondary chart (as each version will potentially have it's own chart).
TLDR; (from my PoV):
- if there is a need for something like separate versions of the extension for operator/seeds, then we need 2 charts (can consider defaulting one `OciRepository` from the other even)
- we should consider 2 separate values because otherwise everything in the runtimeCluster would need to be very close to an actual seed and this is a restriction we don't necessarily want. From the hackathon it was the priorityClasses but I can imaging more things may appear
- otherwise, we go back to the original plan of a single chart/value combo (and we try to bridge the gap between runtime clusters and seeds).


> Why did you deviate from this?

Thought I knew better but I digress. Update coming up

Comment (rfranzke) [reply example/operator/15-extension.yaml]: Aha, makes sense, thanks. Still, I would keep only one `ociRepository` then (i.e., only one version will be used for both runtime and garden deployment).

**Comment by kon-angelo** [comment]
- I added the `deployment` section. 
- Renamed the json fields to simpler names.
- After https://github.com/gardener/gardener/pull/9924#discussion_r1629218977 the `extensionDeployment` now uses a single chart but allows for 2 different values to be used.
- The admission controller was adapted to offer a single `Values` fields to be passed to both charts. This is possible because also these charts in the extension are basically subcharts and passing the values from a single field, should be possible. Not sure if it seems somewhat more complicated.

File: example/operator/15-extension.yaml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+#    type: local
+#    primary: true|false
+#    globallyEnabled: true|false # only valid if kind=Extension
+#    reconcileTimeout: 30s # only valid if kind=Extension
+#    lifecycle: {}
+#    workerlessSupported: true|false # only valid if kind=Extension
+#  deployment:
+#    admission:
+#      values: {}
+#      runtime:
```
Comment (rfranzke) [inline example/operator/15-extension.yaml]:
```suggestion
#      runtimeCluster:
```

File: example/operator/15-extension.yaml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+#        helm:
+#          ociRepository: {}
+#      garden:
```
Comment (rfranzke) [inline example/operator/15-extension.yaml]:
```suggestion
#      virtualCluster:
```

File: example/operator/15-extension.yaml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+#        helm:
+#          ociRepository: {}
+#    extension:
+#      values: {}
+#      runtimeValues: {}
```
Comment (rfranzke) [inline example/operator/15-extension.yaml]:
```suggestion
#      runtimeClusterValues: {}
```

**Review by rfranzke — COMMENTED** [review]
Nits to harmonize this with the wording in the `Garden` API

File: pkg/apis/operator/v1alpha1/types_extension.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+	// Resources is a list of combinations of kinds (DNSRecord, Backupbucket, ...) and their actual types
+	// (aws-route53, gcp).
+	// +optional
+	Resources []gardencorev1beta1.ControllerResource `json:"resources,omitempty"`
+	// Deployment contains deployment configuration for an extension and it's admission controller.
+	// +optional
+	Deployment *Deployment `json:"deployment,omitempty"`
+}
+
+// Deployment specifies how an extension can be installed for a gardener landscape. It includes the specification
```
Comment (MartinWeindel) [inline pkg/apis/operator/v1alpha1/types_extension.go]:
```suggestion
// Deployment specifies how an extension can be installed for a Gardener landscape. It includes the specification
```

File: pkg/apis/operator/v1alpha1/types_extension.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+// for installing an extension and/or an admission controller.
+type Deployment struct {
+	// ExtensionDeployment contains deployment configuration an extension.
```
Comment (MartinWeindel) [inline pkg/apis/operator/v1alpha1/types_extension.go]:
```suggestion
	// ExtensionDeployment contains the deployment configuration an extension.
```

File: pkg/apis/operator/v1alpha1/types_extension.go [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+	// +optional
+	ExtensionDeployment *ExtensionDeploymentSpec `json:"extension,omitempty"`
+	// AdmissionDeployment contains deployment configuration for an admission controller.
```
Comment (MartinWeindel) [inline pkg/apis/operator/v1alpha1/types_extension.go]:
```suggestion
	// AdmissionDeployment contains the deployment configuration for an admission controller.
```

File: pkg/apis/operator/v1alpha1/types_extension.go [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+	// +optional
+	AdmissionDeployment *AdmissionDeploymentSpec `json:"admission,omitempty"`
+}
+
+// ExtensionDeploymentSpec specifies how to install the extension in a gardener landscape. The installation is split into two parts:
+// - installing the extension in the garden cluster by creating the ControllerRegistration and ControllerDeployment.
```
Comment (MartinWeindel) [inline pkg/apis/operator/v1alpha1/types_extension.go]:
```suggestion
// - installing the extension in the virtual garden cluster by creating the ControllerRegistration and ControllerDeployment.
```

File: pkg/apis/operator/v1alpha1/types_extension.go [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+// - installing the extension in the runtime cluster (if necessary).
+type ExtensionDeploymentSpec struct {
+	// DeploymentSpec is the deployment configuration for the extension.
+	// +optional
+	DeploymentSpec `json:",inline"`
+
+	// Values are the deployment values used in the creation of the ControllerDeployment in the garden cluster.
```
Comment (MartinWeindel) [inline pkg/apis/operator/v1alpha1/types_extension.go]:
```suggestion
	// Values are the deployment values used in the creation of the ControllerDeployment in the virtual garden cluster.
```

**Review by MartinWeindel — COMMENTED** [review]
Only some suggestions to improve the documentation.

File: pkg/apis/operator/v1alpha1/types_extension.go [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+// DeploymentSpec is the specification for the deployment of a component.
+type DeploymentSpec struct {
+	// Helm contains the specification for a Helm deployment.
+	Helm *ExtensionHelm `json:"helm,omitempty"`
+}
+
+// ExtensionHelm is the configuration for a helm deployment.
+type ExtensionHelm struct {
+	// OCIRepository defines where to pull the chart.
```
Comment (timuthy) [inline pkg/apis/operator/v1alpha1/types_extension.go]:
```suggestion
	// OCIRepository defines where to pull the chart from.
```

File: pkg/apis/operator/v1alpha1/types_extension.go [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+	// +optional
+	OCIRepository *gardencorev1.OCIRepository `json:"ociRepository,omitempty"`
+}
+
+// ExtensionStatus is the status of a Gardener extension.
+type ExtensionStatus struct {
+	// ObservedGeneration is the most recent generation observed for this resource.
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
```
Comment (timuthy) [inline pkg/apis/operator/v1alpha1/types_extension.go]:
```suggestion
	// ObservedGeneration is the most recent generation observed for this resource.
	// +optional
	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
```

File: pkg/apis/operator/v1alpha1/types_extension.go [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+type ExtensionDeploymentSpec struct {
+	// DeploymentSpec is the deployment configuration for the extension.
+	// +optional
+	DeploymentSpec `json:",inline"`
+
+	// Values are the deployment values used in the creation of the ControllerDeployment in the garden cluster.
+	// +optional
+	Values *apiextensionsv1.JSON `json:"values,omitempty"`
+
+	// RuntimeClusterValues are the deployment values for the extension deployment running in the runtime cluster.
```
Comment (timuthy) [inline pkg/apis/operator/v1alpha1/types_extension.go]:
```suggestion
	// RuntimeClusterValues are the deployment values for the extension deployment running in the runtime garden cluster.
```

File: example/operator/15-extension.yaml [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+#    admission:
+#      values: {}
+#      runtimeCluster:
+#        helm:
+#          ociRepository: {}
+#      virtualCluster:
+#        helm:
+#          ociRepository: {}
+#    extension:
+#      values: {}
```
Comment (timuthy) [inline example/operator/15-extension.yaml:24]:
```suggestion
#      values: {}
#      runtimeClusterValues: {}
```

**Review by timuthy — CHANGES_REQUESTED** [review]
Nice, just a few nits from my side.

File: example/operator/15-extension.yaml (hunk 13 above)
Comment (kon-angelo) [reply example/operator/15-extension.yaml:24]: I am not sure I understand this suggestion. The `# runtimeClusterValues: {}` is the line below right? What am I missing 😅

Comment (timuthy) [reply example/operator/15-extension.yaml:24]: My bad, missed it during review 😃

**Review by timuthy — COMMENTED** [review]
/lgtm

**Review by timuthy — APPROVED** [review]

**Comment by timuthy** [comment]
/hold
To give precedence to #9961.
